src/services/ServerManager.ts

```diff
@@ -0,0 +1,277 @@
+      'DHC',
+      this._configService.getCoreServers()
+    );
+
+    const initialDheServerState = getInitialServerStates(
+      'DHE',
+      this._configService.getEnterpriseServers()
+    );
+
+    this._serverMap = new Map(
+      [...initialDhcServerState, ...initialDheServerState].map(state => [
+        state.url,
+        state,
+      ])
+    );
+
+    // If server config changes in a way that removes servers, disconnect any
+    // active connections from them.
+    for (const serverUrl of this._connectionMap.keys()) {
+      if (!this._serverMap.has(serverUrl)) {
+        this.disconnectFromServer(serverUrl);
+      }
+    }
+  };
+
+  connectToServer = async (serverUrl: URL): Promise<IDhService | null> => {
+    if (this.hasConnection(serverUrl)) {
+      logger.info('Already connected to server:', serverUrl);
+      return null;
+    }
+
+    const serverState = this._serverMap.get(serverUrl);
+
+    // TODO: implement DHE
+    if (serverState == null || serverState.type !== 'DHC') {
+      return null;
+    }
+
+    const connection = this._dhcServiceFactory.create(serverUrl);
+
+    this._connectionMap.set(serverUrl, connection);
+    this._onDidUpdate.fire();
+
+    if (!(await connection.initDh())) {
+      this._connectionMap.delete(serverUrl);
+    }
+
+    this._onDidConnect.fire(serverUrl);
+    this._onDidUpdate.fire();
+
+    return this._connectionMap.get(serverUrl) ?? null;
+  };
+
+  disconnectEditor = (uri: vscode.Uri): void => {
+    this._uriConnectionsMap.delete(uri);
+    this._onDidUpdate.fire();
+  };
+
+  disconnectFromServer = async (serverUrl: URL): Promise<void> => {
+    const connection = this._connectionMap.get(serverUrl);
+
+    if (connection == null) {
+      return;
+    }
+
+    this._connectionMap.delete(serverUrl);
+
+    // Remove any editor URIs associated with this connection
+    this._uriConnectionsMap.forEach((dhService, uri) => {
+      if (dhService === connection) {
+        this._uriConnectionsMap.delete(uri);
+      }
+    });
+
+    await connection.dispose();
+
+    this._onDidDisconnect.fire(serverUrl);
+    this._onDidUpdate.fire();
+  };
+
+  /**
+   * Determine if the given server URL has any active connections.
+   * @param serverUrl
+   */
+  hasConnection = (serverUrl: URL): boolean => {
+    return this._connectionMap.has(serverUrl);
+  };
+
+  /**
+   * Determine if the given connection is assicated with any editor URIs.
+   * @param connection
+   */
+  hasConnectionUris = (connection: IDhService): boolean => {
+    for (const cn of this._uriConnectionsMap.values()) {
+      if (cn === connection) {
+        return true;
+      }
+    }
+
+    return false;
+  };
+
+  getServers = ({
+    isRunning,
+    hasConnections,
+  }: {
+    isRunning?: boolean;
+    hasConnections?: boolean;
+  } = {}): ServerState[] => {
+    // Start polling server status the first time servers are requested.
+    // TBD: Is there a way to stop this when the servers list goes out of view?
```

Comment:
We absolutely need a way to stop polling when the extension is not in view... or have it simply be triggered by events. Having an extension polling every 3s without a way to stop it is a reason for someone to not install this. Also it's checking both disconnected and connected servers every 3s.
We could probably be smart and get away with polling entirely if we just refresh when the user is interacting with Deephaven:
- When the sidebar is activated, check all servers
- When a Python file is focused that is associated with a Deephaven connection, check if it's connected
- Check when user clicks the bottom bar
- An explicit refresh button in the Deephaven extension bar
- DHE: Listen on the `Client` object for disconnect/reconnect after we're connected, so we don't need to poll for disconnect at all (or really be checking at this level, since it's built-in to [IrisClient](https://github.com/deephaven-ent/iris/blob/d802ba4f870e4a2dd19a6af61e1ac788a0caaf9e/web/client-api/src/main/java/com/illumon/iris/web/client/api/IrisClient.java#L81), using exponential backoff).

As it is right now polling every 3 seconds and retrieving the JS API is a lot. The file is 2MB and is not cached.

Reply:
I found a way to do this. Now only polls when window is active + focused and the server treeview is visible. It manually runs 1x whenever server config is loaded.